Makefile

```diff
@@ -188,8 +188,48 @@ TF_CFLAGS_aarch64	+=	-mgeneral-regs-only -mstrict-align
 ASFLAGS_aarch32		=	$(march32-directive)
 ASFLAGS_aarch64		=	-march=armv8-a
 
+WARNING1 := -Wextra
+WARNING1 += -Wunused -Wno-unused-parameter
+WARNING1 += -Wmissing-declarations
+WARNING1 += -Wmissing-format-attribute
+WARNING1 += -Wmissing-prototypes
+WARNING1 += -Wold-style-definition
+WARNING1 += -Wunused-but-set-variable
+WARNING1 += -Wunused-const-variable
+
+WARNING2 := -Waggregate-return
+WARNING2 += -Wcast-align
+WARNING2 += -Wdisabled-optimization
+WARNING2 += -Wnested-externs
+WARNING2 += -Wshadow
+WARNING2 += -Wlogical-op
+WARNING2 += -Wmissing-field-initializers
+WARNING2 += -Wsign-compare
+WARNING2 += -Wmaybe-uninitialized
+
+WARNING3 := -Wbad-function-cast
+WARNING3 += -Wcast-qual
+WARNING3 += -Wconversion
+WARNING3 += -Wpacked
+WARNING3 += -Wpadded
+WARNING3 += -Wpointer-arith
+WARNING3 += -Wredundant-decls
+WARNING3 += -Wswitch-default
+WARNING3 += -Wpacked-bitfield-compat
+WARNING3 += -Wvla
+
+ifeq (${W},1)
+WARN_OR_ERROR := $(WARNING1)
+else ifeq (${W},2)
+WARN_OR_ERROR := $(WARNING1) $(WARNING2)
+else ifeq (${W},3)
+WARN_OR_ERROR := $(WARNING1) $(WARNING2) $(WARNING3)
+else
+WARN_OR_ERROR := -Werror
+endif
+
 CPPFLAGS		=	${DEFINES} ${INCLUDES} ${MBEDTLS_INC} -nostdinc		\
-				-Wmissing-include-dirs -Werror
+				-Wmissing-include-dirs $(WARN_OR_ERROR)
```

Comment:
It's a bit weird to leave ``-Wall`` in ``TF_CFLAGS``, right?

Reply:
With this patch I didn't want to change what is currently done, that's why I kept the -Werror if no W=x option is provided in the command line.
-Wall may be misleading, as it does not includes the warnings that are added in this patch, as stated here: https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html

Reply:
I think it would be useful to set W=3 and also -Werror, once all the warning are fixed. In this case, it would be better to disable -Werror with another variable and have it enabled by default.

Reply:
OK, I can then keep the -Werror as it is now, just add the warnings if W=x is provided. And then make another patch to put -Werror depending on a command line variable, why not E=0 to deactivate errors?

Reply:
OK, sounds good.